Add option to have gpg decrypt failures treated as errors

jfindlay opened the pull request:
### What does this PR do?

This is defaulted `false` for compatibility, and is controlled via a `gpg_decrypt_must_succeed` option.


### Previous Behavior

If this is `True`- and the ciphertext couldn't be decrypted- then it's treated as an error rather than just passing the raw ciphertext through.

Sending the ciphertext through basically is *never* desired- the point of the gpg renderer is to decrypt the secret, not send the cipher text through.

### New Behavior

Explicitly error if gpg does not decrypt the ciphertext.

### Merge requirements satisfied?
- [x] Docs
- [x] Changelog - https://docs.saltproject.io/en/master/topics/development/changelog.html
- [x] Tests written/updated

### Commits signed with GPG?
Yes

Commit:
Add option to have gpg decrypt failures treated as errors.

This is defaulted `false` for compatibility, and is controlled via
a `gpg_decrypt_must_succeed` option.

If this is true- and the ciphertext couldn't be decrypted- then it's
treated as an error rather than just passing the raw ciphertext through.

Sending the ciphertext through basically is *never* desired- the point
of the gpg renderer is to decrypt the secret, not send the cipher text
through.

jfindlay committed:
Add tests for gpg decryption failure option

Test that:
1. Pillar registers an error when `gpg_decrypt_must_succeed` is `True` and decryption fails
2. The GPG renderer fails silently when `gpg_decrypt_must_succeed` is `False`

Also mock `__opts__["gpg_decrypt_must_succeed"]` for gpg renderer unit pytests.

frebib commented on salt/renderers/gpg.py:
```diff
@@ -431,6 +453,12 @@ def render(gpg_data, saltenv="base", sls="", argline="", **kwargs):
         raise SaltRenderError("GPG unavailable")
     log.debug("Reading GPG keys from: %s", _get_key_dir())
 
+    if not __opts__["gpg_decrypt_must_succeed"]:
```
Sane defaults 🙏
Nice

twangboy replied:
Maybe
```
if not __opts__.get("gpg_decrypt_must_succeed"):

twangboy replied:
That's the error I'm seeing in tests... key does not exist...

jfindlay replied:
right, but I would rather the config be expected to be there.  Since it's defined in `DEFAULT_MASTER_OPTS`, something in the test suite might be overwriting or modifying `__opts__`, in which case the `KeyError` is a sign of a deeper problem.

jfindlay replied:
@twangboy I think I misread earlier that the errors were coming from unit tests, not the functional tests.  All other lookups of `__opts__` in `salt/renderers/gpg.py` go through `.get()`, which in my opinion is more dubious than a clean fail if a necessary option is not available, thus the [`configure_loader_modules`](https://docs.saltproject.io/en/latest/topics/development/tests/unit.html#mocking-loader-modules) [comment](https://github.com/saltstack/salt/pull/61418/#issuecomment-1015874055) from @MKLeb .

MKLeb commented on tests/pytests/functional/pillar/test_gpg.py:
```diff
@@ -184,14 +184,34 @@
 }
 
+# Random data pretending to be ciphertext
+GPG_PILLAR_YAML_FAIL = """\
+fail: |
+  -----BEGIN PGP MESSAGE-----
+
+  OzyJmQJJVPxGQqyxwIcAl0wWqdSTHpMXYPLrDoRU8H1xa2DhE5DeUihjm4fHUcHp
+  -----END PGP MESSAGE-----
+"""
+
+GPG_PILLAR_ENCRYPTED_FAIL = {
+    "fail": (
+        "-----BEGIN PGP MESSAGE-----\n"
+        "\n"
+        "OzyJmQJJVPxGQqyxwIcAl0wWqdSTHpMXYPLrDoRU8H1xa2DhE5DeUihjm4fHUcHp\n"
+        "-----END PGP MESSAGE-----\n"
+    ),
+}
+
```
Would you mind turning these (and the other globals in this file) into pytest fixtures?

MKLeb commented:
Something like the following in `tests/pytest/unit/renderers/test_gpg.py` may help with test failures...

```
@pytest.fixture
def configure_loader_modules():
    """
    GPG renderer configuration
    """
    return {
        gpg: {
            "__opts__": {
                "gpg_decrypt_must_succeed": True
            }
        }
    }
```

jfindlay committed:
Convert gpg pytest constants to fixtures

- Normalize module-global data into pytest fixtures.  It is not the
  least opaque option but we work within the constraints of pytest.
- Also rescope gpg pillar fixtures to module.  If they are needed in a
  larger context, they should be defined in an appropriately prominent
  file.

Ch3LL commented on salt/renderers/gpg.py:
```diff
@@ -431,6 +453,12 @@ def render(gpg_data, saltenv="base", sls="", argline="", **kwargs):
         raise SaltRenderError("GPG unavailable")
     log.debug("Reading GPG keys from: %s", _get_key_dir())
 
+    if not __opts__["gpg_decrypt_must_succeed"]:
```
I'm worried that adding a deprecation warning whenever the files are rendered would be too noisy. Can we add the deprecation notice when its not able to decrypt?

jfindlay commented:
@Ch3LL anything else that needs to be done on this?

Ch3LL commented:
Apologies for the delay, I was out sick most of last week. Thanks it looks good :)

jfindlay commented:
> Apologies for the delay, I was out sick most of last week. Thanks it looks good :)

No problem, thanks for the review.

max-arnold commented:
I guess it is too late to fix, but Salt has two configuration namespaces for feature toggles:

1. [`features`](https://docs.saltproject.io/en/latest/ref/states/vars.html)
2. `use_superseded`

This means that instead of a global `gpg_decrypt_must_succeed` option, the toggle could be implemented as:


```
features:
  strict_gpg_decrypt
```

```
from salt.features import features
if features.get("strict_gpg_decrypt", False):
    ...
```

or

```
use_superseded:
  - module.foo
```

```
from salt.utils.decorators import with_deprecated

@with_deprecated(globals(), "Beryllium")
def foo():
    "This is a new function"

def _foo():
    "This is a deprecated function"
```

P.S. It would be nice to deprecate the non-strict way in future.  These feature flags tend to accumulate in various master/minion bootstrap scripts I'm working on. I'm always torn between enabling better defaults (`module.run`, `enable_slsvars_fixes`, `enable_legacy_startup_events: false`, etc) and simpler bootstrap scripts.

jfindlay commented:
> It would be nice to deprecate the non-strict way in future (or flip the default).
This was my intention.  It's clearly the obvious thing to do.  Let the option bake a couple of releases and then there's no reason to change the default behavior.  You're welcome to submit the pull request, otherwise I will get around to actioning it from my backlog at some point.
